[Frontend] Add unix domain socket support

yyweiss opened the pull request:
Adds support for serving vllm using unix sockets instead of network ports.

I wasn't sure exactly where to add documentation about the feature. I'll try again later, but would appreciate some pointers.

The changes should not change the behavior of anything which previously worked. I'll try to set up a testing environment, and add some tests for this feature.

Other suggestions or corrections are also welcome.

FIX #13907

yyweiss committed:
Add --uds command line argument to vllm serve

Add an argument `--uds` to the `vllm serve` argument parser
intended to make the server available over a Unix domain socket
instead of a network port.

(updated to new config format)

Signed-off-by: yyweiss <yyweiss@gmail.com>

Signed-off-by:  <yyweiss@gmail.com>
Signed-off-by: yyw <yyweiss@gmail.com>

yyweiss committed:
New function to create a socket bound to uds

Added a function `create_server_unix_socket`. Can be merged into
`create_server_socket` instead by adding a parameter for socket type.

(edited to resolve merge conflict)

Signed-off-by: yyweiss <yyweiss@gmail.com>

Signed-off-by:  <yyweiss@gmail.com>
Signed-off-by: yyw <yyweiss@gmail.com>

yyweiss committed:
Call correct function for socket creation

`create_server_socket` for network socket.
`create_server_unix_socket` for unix socket.

Signed-off-by: yyweiss <yyweiss@gmail.com>

Signed-off-by:  <yyweiss@gmail.com>
Signed-off-by: yyw <yyweiss@gmail.com>

russellb commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -1080,6 +1080,15 @@ def create_server_socket(addr: tuple[str, int]) -> socket.socket:
     return sock
 
 
+def create_server_unix_socket(addr: str) -> socket.socket:
+    family = socket.AF_UNIX
+
+    sock = socket.socket(family=family, type=socket.SOCK_STREAM)
+    sock.bind(addr)
```
```suggestion
def create_server_unix_socket(path: str) -> socket.socket:
    sock = socket.socket(family=socket.AF_UNIX, type=socket.SOCK_STREAM)
    sock.bind(path)
```

russellb commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -1080,6 +1080,15 @@ def create_server_socket(addr: tuple[str, int]) -> socket.socket:
+
+    sock = socket.socket(family=family, type=socket.SOCK_STREAM)
+    sock.bind(addr)
+
+    return sock
```
```suggestion
    return sock
```

russellb commented on vllm/entrypoints/openai/cli_args.py:
```diff
@@ -86,6 +86,10 @@ def make_arg_parser(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
+    parser.add_argument("--uds",
+                        type=optional_type(str),
+                        default=None,
+                        help="Unix domain socket name.")
```
Some more points to cover here:

- If used, it takes precedence over the options for setting up a TCP or TCP+SSL socket for the API server. Only one or the other will be used, not both.

russellb reviewed (changes requested):
Can you also take a look in `docs/source/` to find an appropriate place to document the feature there as well?

yyweiss committed:
Style cleanup from following review

Signed-off-by: yyw <yyweiss@gmail.com>

Signed-off-by:  <yyweiss@gmail.com>
Signed-off-by: yyw <yyweiss@gmail.com>

yyweiss committed:
Clarify UDS precedence over TCP

(Indentation change done by precommit hook)

Signed-off-by: yyw <yyweiss@gmail.com>

Signed-off-by:  <yyweiss@gmail.com>
Signed-off-by: yyw <yyweiss@gmail.com>

yyweiss commented:
Thanks.

1. I modified the code the way you suggested.
2. I added a note to the `--uds` argument about the precedence.
3. If I understand correctly, the code modification will automatically update <https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html#cli-reference> which seems to be the appropriate location.

yyweiss committed:
Edit log message for server start on unix socket

Creates a "listen_address" string for a unix socket server.
Added following merge conflicts.

Signed-off-by: yyw <yyweiss@gmail.com>

Signed-off-by:  <yyweiss@gmail.com>
Signed-off-by: yyw <yyweiss@gmail.com>

russellb commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -1313,8 +1323,10 @@ def signal_handler(*_) -> None:
     host_part = f"[{addr}]" if is_valid_ipv6_address(
         addr) else addr or "0.0.0.0"
-    listen_address = f"http{'s' if is_ssl else ''}://{host_part}:{port}"
-
+    if args.uds:
+        listen_address = f"unix:{args.uds}"
+    else:
+        listen_address = f"http{'s' if is_ssl else ''}://{host_part}:{port}"
```
The lines starting with `addr, port = sock_addr` will need to be moved under the `else:` block here, since they are non-uds specific.

yyweiss replied:
Done

russellb reviewed (changes requested):
some minor refactoring suggestions.

It would also be great if you could cover this somewhere in `docs/`.

yyweiss committed:
Do not use sock_addr when using uds

Do not create the unused sock_addr=(host,port) when using
a unix socket.

Signed-off-by: yyw <yyweiss@gmail.com>

Signed-off-by:  <yyweiss@gmail.com>
Signed-off-by: yyw <yyweiss@gmail.com>

yyweiss committed:
Add vllm serve example using UDS

(fixed to match new indentation)

Signed-off-by: yyw <yyweiss@gmail.com>

Signed-off-by:  <yyweiss@gmail.com>
Signed-off-by: yyw <yyweiss@gmail.com>

yyweiss commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -1296,8 +1302,12 @@ def setup_server(args):
-    sock_addr = (args.host or "", args.port)
-    sock = create_server_socket(sock_addr)
+    if args.uds:
+        sock_addr = (args.uds, args.port)
```
I completely removed line 1306

russellb reviewed (commented):
Thanks!  One last request - can you look at adding a functional test for this to make sure it stays functional?

yyweiss commented:
@russellb 
I made the changes you suggested.

I added an example of running `vllm serve` over a unix socket to https://docs.vllm.ai/en/latest/cli/index.html#serve

Documentation of the command line arguments are created automatically, so I don't think I need to do that myself.

Looking at the documentation, I should probably also add (usage) support to `vllm chat` and `vllm completions`, but that could be a separate issue.

yyweiss commented:
My previous comment was written before I saw the additional request for a test.

I'll copy one of the basic tests and modify the arguments to use a Unix socket. 

It looks like the testing requirements already include httpx, so I'll use that for making the request.

I'll also need to modify the RemoteOpenAIServer to provide an appropriate base url. That should be enough for the my test, though I'm not sure if the openai client returned by get_{,async_}client  will work.

I'm also debating if to add a named argument for the initializer, or to just look at the arguments. @russellb What do you think?

russellb commented:
@yyweiss hey, apologies that I didn't answer sooner. I don't have a strong opinion regarding your question. Did you get a chance to start on the test for this?

yyweiss commented:
Thanks for the reminder. I started writing the test, and then got distracted.

I found that in addition to copying and modifying an existing test (which is simple and safe enough), I'd also have to modify the test client.

Since this involves touching code which can potentially affect - and if I do something wrong, break - existing test code, I would prefer to have this PR merged, and to then open a new issue for the test, if that is okay with you.

yyweiss committed:
Add test for UDS support attempt 1

- Copied and modified one of the existing basic tests.

- Test uses httpx instead of requests.

- Modified the test RemoteOpenAIServer to allow its use with UDS.

Signed-off-by: yyw <yyweiss@gmail.com>

Signed-off-by:  <yyweiss@gmail.com>
Signed-off-by: yyw <yyweiss@gmail.com>

yyweiss committed:
Fixed a missed merge conflict

Signed-off-by: yyw <yyweiss@gmail.com>

Signed-off-by:  <yyweiss@gmail.com>

yyweiss committed:
Fix SPDX header

For some reason, the local pre-commit hook doesn't check that

Signed-off-by: yyw <yyweiss@gmail.com>

Signed-off-by:  <yyweiss@gmail.com>

yyweiss commented:
I created a test.

I think the entrypoint/openai server tests completed successfully, so I probably didn't break the existing tests.

@russellb

yuikns commented:
@russellb   Just a gentle nudge—anything still blocking this? I'd really love to use Unix-socket support.

russellb reviewed (approved):
Apologies for the long delay. I had some vacation and other things limiting time I could spend on reviews.

This LGTM. Thanks for the work!
